doc/architecture.md

```diff
@@ -0,0 +1,206 @@
+  while `DandiDav::propfind()` always returns a "multistatus" WebDAV XML
+  document.
+
+    - When rendering information about blob assets (either on their own or
+      within a parent resource), a choice must be made about what URL to
+      provide as the download URL.  The `contentUrl` metadata field of a blob
+      asset is assumed to contain two download URLs, one pointing to S3 and one
+      pointing to an Archive API endpoint that redirects to a signed version of
+      the S3 URL that sets a filename in the `Content-Disposition` header.  The
+      latter type of URL is generally more desirable, as it means that any
+      users downloading the URL via a browser will end up with a file with the
+      same filename as the asset, while a direct, unsigned S3 URL will be
+      downloaded as a file with the same filename as the S3 key, which is
+      usually the blob ID and not user-friendly.  Hence, by default, a `GET`
+      request to `dandidav` for a blob asset will be responded to with a
+      redirect to the Archive API download URL.  Unfortunately, certain WebDAV
+      clients (i.e., [davfs2](https://savannah.nongnu.org/bugs/?65376)) do not
+      support WebDAV servers that return redirects to other redirects, so the
+      `--prefer-s3-redirects` CLI option was added to `dandidav` to instead
+      make these `GET` requests redirect directly to the unsigned S3 URLs.
+
+        - Note that HTML listings of a collection's children will always link
+          blob assets to their Archive API download URLs, regardless of
+          `--prefer-s3-redirects`, as these listings are only returned for
+          `GET` requests to a collection, which WebDAV clients don't do.
+          (These listings link blob assets and Zarr entries directly to
+          download URLs rather than `dandidav` resource URLs because doing the
+          latter would result in subsequent requests to `dandidav` that just
+          duplicate work.)
+
+        - Note that this does not apply to download URLs for Zarr entries, as
+          their S3 URLs are assumed to always have the same filename as the
+          entry, and thus signing to add a `Content-Disposition` is
+          unnecessary.
+
+- [`DandiDav::get_resource()`][] and
+  [`DandiDav::get_resource_with_children()`][] perform method calls on a
+  `DandiClient` or `ZarrManClient` instance (as appropriate) belonging to the
+  `DandiDav` instance in order to fetch information about the specified
+  resource(s).
+
+
+`DandiClient`
+-------------
+
+- The [`DandiClient`][] type is used to retrieve information about resources on
+  an Archive instance (the one specified via the `--api-url` option on program
+  invocation) via a combination of the Archive's API and (for entries & folders
+  inside Zarrs) requests to an S3 bucket.  It is the data source for the
+  `/dandisets/` hierarchy served by `dandidav`.
+
+    - Information about child resources of Zarrs needs to be fetched via S3
```

Comment:
don't we use manifest json files here from https://datasets.datalad.org/dandi/zarr-manifests/ and not by traversing S3 or our API server ATM?

Reply:
The Zarr manifests are used by the `/zarrs/` hierarchy.  The `/dandisets/` hierarchy makes direct requests to S3.